text/0813-template-layering.md

```diff
@@ -0,0 +1,984 @@
+export – `template`, which is a re-export of the `template()` function we
+described above. It has the same semantics as `template()` from the main module
+except it is always invalid to use this imported function as a tag, so linters
+and TypeScript can trivial flag this invalid usage.
+
+The purpose of the re-exported `template()` from the runtime module is to
+signal to both human readers and build tools that the template compilation is
+deliberately deferred to runtime. These usages must not be processed at build
+time and linters must not warn about non-idiomatic usages.
+
+Typically, developers should favor importing `template()` from the main module
+so that their code can be agonistic against where the compilation is happening.
+However, since the intent is that build-time compilation should be _possible_,
+they should also take care to ensure they stay within the set of idiomatic set
+of usages to guarantee that is the case, and this re-export from the runtime
+module offers a way to signal that the deviations are deliberate.
+
+This RFC does not propose to add re-exports for other features found in the
+`@ember/template-compilation` package. Conceptually, it makes sense to provide
+the same ex-export for at least the `compile()` function. However, as we move
+to a world where components are used everywhere and the need for "standalone
+templates" dwindles, it is unclear that the `compile()` function is needed in
+the long run. In any case, if that turns out to be necessary, a future RFC can
+propose to add support for that and the precedence set forth in this RFC should
+make that a straightforward process.
+
+### TypeScript
+
+We don't expect this feature to pose any significant challenges in terms of
+TypeScript support, other than having a few overloaded signatures/purpose on
+the same `template()` function may make things a little cumbersome, but not
+prohibitively so. The signature of the function can be described as such:
+
+```ts
+// These types are defined in other RFCs
+interface Signature {
+  /* ... */
+}
+interface Component<S extends Signature> {
+  /* ... */
+}
+
+type Scope = () => Record<string, unknown>;
+
+function template<S extends Signature>(source: string): Component<S>;
+function template<S extends Signature>(
+  source: string,
+  scope: Scope
+): Component<S>;
+function template<S extends Signature, C extends Component<S>>(
+  source: string,
+  scope: Scope,
+  target: C
+): C;
+
+// For the build-time tagged template literal extension. By not specifying
+// additional arguments here, TypeScript actually enforces that the tagged
+// string literal cannot have any interpolation expressions.
+function template<S extends Signature>(
+  source: TemplateStringsArray
+): Component<S>;
+```
+
+One nice thing about this design is that there is a natural place for the type
+(generic) argument for the template's signature, negating the need for the
+special `<template[Signature]>` syntax.
```

Comment:
This actually does *not* solve the issue from a general authoring perspective, though, I think? After all, we're not going to be teaching people to write it this way in general, and we’re *definitely* not going to be implementing Glint support for it.

Reply:
Yep, it should probably say it gives you a place to emit into. But what I was trying to say (badly) is that it does not require specifically designing something in this API that does that equivalent of `<template[Signature]>` because it already naturally exists

Reply:
Ah, yep, that’s clearer and I agree: We still need surface syntax for it, but it’s nice that there’s a natural home for it in the compile target. 👍🏼